### DISCUSSION TITLE
internal server error

### DISCUSSION THREAD

**Discussion by cibiman** [evroon/bracket#391] [Q&A]
New installation and error when trying to log in
![error](https://github.com/evroon/bracket/assets/122181788/9939e520-e815-40b3-a746-4e3581d83e32)

**Comment by evroon** [comment]
Hi, how do you run this? (via Docker, or by running it [without Docker](https://github.com/evroon/bracket/#running-the-frontend-and-backend-for-development) example)
Most likely the frontend can't connect to the backend.
I just did what is written in [quickstart](https://github.com/evroon/bracket/#quickstart) and for me it works correctly.

**Comment by cibiman** [comment]
Hi, I have use Quickstart . 3 containers are up and running , no errors in log. You are right, when I stop the bracket-backend I get the same error. Why there are no communication? They are in the same network. I am not an expert in docker.

**Comment by evroon** [comment]
can you maybe run `sudo docker-compose logs` and look for errors?
There should be lines like these:
```
bracket-backend     | [2023-12-08 16:19:15 +0000] [1] [INFO] Starting gunicorn 21.2.0
bracket-backend     | [2023-12-08 16:19:15 +0000] [1] [INFO] Listening at: http://0.0.0.0:8400 (1)
```
But you might see an error instead of the above.

For the frontend, you probably do see this:
```
bracket-frontend    | ready - started server on 0.0.0.0:3000, url: http://localhost:3000
```

Also, you might have bracket running twice, which blocks one of the backend instances to use port 8400

**Comment by cibiman** [comment]
[2023-12-08 16:05:55 +0000] [1] [INFO] Starting gunicorn 21.2.0
bracket-backend     | [2023-12-08 16:05:55 +0000] [1] [INFO] Listening at: http://0.0.0.0:8400 (1)
bracket-backend     | [2023-12-08 16:05:55 +0000] [1] [INFO] Using worker: uvicorn.workers.UvicornWorker
bracket-backend     | [2023-12-08 16:05:55 +0000] [9] [INFO] Booting worker with pid: 9
bracket-backend     | [2023-12-08 16:05:55 +0000] [10] [INFO] Booting worker with pid: 10
bracket-backend     | [2023-12-08 16:05:56,292] [bracket] [INFO] Current env: DEVELOPMENT
bracket-backend     | [2023-12-08 16:05:56,328] [bracket] [INFO] Current env: DEVELOPMENT
bracket-backend     | [2023-12-08 16:05:57 +0000] [9] [INFO] Started server process [9]
bracket-backend     | [2023-12-08 16:05:57 +0000] [9] [INFO] Waiting for application startup.
bracket-backend     | [2023-12-08 16:05:57 +0000] [10] [INFO] Started server process [10]
bracket-backend     | [2023-12-08 16:05:57 +0000] [10] [INFO] Waiting for application startup.
bracket-backend     | [2023-12-08 16:05:57 +0000] [9] [INFO] Application startup complete.
bracket-backend     | [2023-12-08 16:05:57 +0000] [10] [INFO] Application startup complete.


bracket-frontend    | yarn run v1.22.19
bracket-frontend    | $ next dev
bracket-frontend    |    ▲ Next.js 14.0.3
bracket-frontend    |    - Local:        http://localhost:3000
bracket-frontend    |
bracket-frontend    |  ✓ Ready in 4.4s
bracket-frontend    |  ○ Compiling /login ...
bracket-frontend    |  ✓ Compiled /login in 6.7s (1093 modules)
bracket-frontend    |  ○ Compiling / ...
bracket-frontend    |  ✓ Compiled / in 1830ms (1268 modules)

**Comment by evroon** [comment]
Okay, and can you open the console in the browser (pressing F12) and make a screenshot of what happens there under the `Networks` tab? There's either 500 status code responses or just failed requests that don't connect to the backend.

I'll make this a discussion btw.

**Comment by cibiman** [comment]
![console](https://github.com/evroon/bracket/assets/122181788/153c2d1a-57b2-48c4-a044-a277c405d8a9)

**Comment by adamheath** [comment]
I am facing the same issue when using the docker quickstart instructions. Has there been any progress on this?

**Comment by evroon** [comment]
I think the problem is because of CORS, which I have now disabled by default: https://github.com/evroon/bracket/pull/400
Could you try again with the latest version of Bracket?
Another problem could be the port forwarding in Docker that you might have misconfigured, but it should be fine if you just use this one without adjustment: https://github.com/evroon/bracket/blob/master/docker-compose.yml

**Comment by adamheath** [comment]
Still the same issue. Steps taken:

git clone https://github.com/evroon/bracket.git
cd bracket
sudo docker-compose up -d
sudo docker exec bracket-backend pipenv run ./cli.py create-dev-db

When looking at the network inspect tab it seems like it is trying to connect to localhost:8400 rather than the docker backend IP although my knowledge on all this is limited

<img width="545" alt="image" src="https://github.com/evroon/bracket/assets/548417/26866142-d1b0-4653-a14e-932fc567bdcd">

**Reply by cibiman** [reply]
I saw that too. Only works locally.

**Reply by evroon** [reply]
[This line](https://github.com/evroon/bracket/blob/master/docker-compose.yml#L15) forwards the port 8400 from the docker container to localhost, so that's why localhost:8400 should just work.
I think there's a few things you can check:
- Do you have some kind of firewall (e.g. ufw) enabled for port 8400? 
- What does the `Response` tab in the console say for these requests?
- Can you at least run `curl localhost:8400/ping`?
- Does `sudo docker ps -a` show something like the following? (note the PORTS column in particular)

```
CONTAINER ID   IMAGE                             COMMAND                  CREATED         STATUS                    PORTS                                       NAMES
88af2bcf1880   ghcr.io/evroon/bracket-backend    "/bin/sh -c 'pipenv …"   3 minutes ago   Up 2 minutes              0.0.0.0:8400->8400/tcp, :::8400->8400/tcp   bracket-backend
be1306b8167c   postgres                          "docker-entrypoint.s…"   3 minutes ago   Up 3 minutes              5432/tcp                                    bracket_postgres_1
987f1194605b   ghcr.io/evroon/bracket-frontend   "docker-entrypoint.s…"   3 minutes ago   Up 3 minutes              0.0.0.0:3000->3000/tcp, :::3000->3000/tcp   bracket-frontend
```

**Reply by cibiman** [reply]
When run on localhost it works. When I try remotely, it doesn't work. Because the backend does not indicate the IP, only localhost.

**Reply by evroon** [reply]
oh I didn't get that your backend runs on a different host than your browser/frontend.
Can't you use a reverse proxy like NGINX for this?

**Reply by cibiman** [reply]
I installed it on the server and ran it remotely.
 Why does the frontend work remotely? I use traefik.

**Reply by evroon** [reply]
ohh but what happens if you try to access the backend remotely without frontend? 
So just open `http://<backend host ip>:8400/docs`

**Reply by evroon** [reply]
I think the problem is just that the frontend doesn't know where the backend is running so you have to follow the instructions here:
https://github.com/evroon/bracket/issues/326#issuecomment-1804479552

I will try to write it in the documentation soon

**Reply by adamheath** [reply]
in what scenario would running it all locally be beneficial? surely only localhost would be able to login and so you wouldn't be able to have any players access the site?

I have tried following those instructions but my request still tries to go to localhost rather than the remote IP I set.

I'll just wait for the documentation update as I'm clearly missing something

**Reply by homa8511** [reply]
Did I get it right that the bracket-backend needs its own endpoint like api.bracket.mydomain.com that must differ from frontend endpoint?

Currently I receive 502 bad gateway accessing bracket-frontend behind traefik reverse proxy. 

Frontend, backend and database are inside their own docker network and should be able to access each other. But only frontend is exposed to internet by external docker network.

No errors found in traefik access log or in bracket log files. Maybe debugging can be enabled to obtain more information?!

**Reply by evroon** [reply]
Yes, the backend needs to run on a different host/port combination than the frontend, and the frontend needs to know where the backend can be accessed via the `NEXT_PUBLIC_API_BASE_URL` environment variable

**Reply by homa8511** [reply]
Even though I find it strange that a backend is on the Internet, I create a corresponding Traefik configuration for the bracket backend and try it again.

Wouldn't it be possible for the API url to simply use the container name from the backend, including the port?

I was able to set up other Docker applications such as Mailcow or Nextcloud according to this scheme.

**Reply by evroon** [reply]
A backend always has to be exposed to the internet. The frontend runs in your browser and sends requests to a backend running on a server, that's how any web application with javascript (so non-static) works.

> Wouldn't it be possible for the API url to simply use the container name from the backend, including the port?

That's a special case that only works locally, it's no different from exposing port 8400 of the backend and then setting `NEXT_PUBLIC_API_BASE_URL: "http://localhost:8400"`

**Reply by homa8511** [reply]
Docker spins up the backend on a different ip. So localhost won't work. I will try 172.x.0.y as API endpoint. Also bracket-backend container name and last but not least enable backend to be accessed by a public url.

EDIT
still http 502 accessing bracket frontend.

Tried the following regarding NEXT_PUBLIC_API_BASE_URL:

- internal docker network ip address
- bracket-backend container's name
- public uri with traefik reverse proxy (api.bracket.mydomain.com)
- - http 502 on api uri, too. But some assets seems to be loaded like a png file (stated in network tab in browser console)

**Comment by K0styan** [comment]
Hello!
I have caught the same error.
In logs I found the following error messages from backend:

bracket-backend   | [2024-02-08 13:06:39,102] [bracket] [WARNING] Empty db detected, creating admin user...
bracket-backend   | [2024-02-08 13:06:39 +0000] [10] **[ERROR] Traceback (most recent call last):**
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
bracket-backend   |     self.dialect.do_execute(
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
bracket-backend   |     cursor.execute(statement, parameters)
bracket-backend   | psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "pg_type_typname_nsp_index"
bracket-backend   | DETAIL:  Key (typname, typnamespace)=(stage_type, 2200) already exists.
bracket-backend   |
bracket-backend   |
bracket-backend   | The above exception was the direct cause of the following exception:
bracket-backend   |
bracket-backend   | Traceback (most recent call last):
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/starlette/routing.py", line 738, in lifespan
bracket-backend   |     async with self.lifespan_context(app) as maybe_state:
bracket-backend   |   File "/usr/local/lib/python3.10/contextlib.py", line 199, in __aenter__
bracket-backend   |     return await anext(self.gen)
bracket-backend   |   File "/app/bracket/app.py", line 40, in lifespan
bracket-backend   |     await init_db_when_empty()
bracket-backend   |   File "/app/bracket/utils/db_init.py", line 102, in init_db_when_empty
bracket-backend   |     metadata.create_all(engine)
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/sql/schema.py", line 4930, in create_all
bracket-backend   |     bind._run_ddl_visitor(
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3238, in _run_ddl_visitor
bracket-backend   |     conn._run_ddl_visitor(visitorcallable, element, **kwargs)
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2221, in _run_ddl_visitor
bracket-backend   |     visitorcallable(self.dialect, self, **kwargs).traverse_single(element)
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/sql/visitors.py", line 524, in traverse_single
bracket-backend   |     return meth(obj, **kw)
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/sql/ddl.py", line 842, in visit_metadata
bracket-backend   |     metadata.dispatch.before_create(
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/event/attr.py", line 346, in __call__
bracket-backend   |     fn(*args, **kw)
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 875, in __call__
bracket-backend   |     return getattr(self.target, self.name)(*arg, **kw)
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/sql/sqltypes.py", line 1205, in _on_metadata_create
bracket-backend   |     t._on_metadata_create(target, bind, **kw)
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/dialects/postgresql/base.py", line 2218, in _on_metadat>
bracket-backend   |     self.create(bind=bind, checkfirst=checkfirst)
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/dialects/postgresql/base.py", line 2111, in create
bracket-backend   |     bind._run_ddl_visitor(self.EnumGenerator, self, checkfirst=checkfirst)
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2221, in _run_ddl_visitor
bracket-backend   |     visitorcallable(self.dialect, self, **kwargs).traverse_single(element)
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/sql/visitors.py", line 524, in traverse_single
bracket-backend   |     return meth(obj, **kw)
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/dialects/postgresql/base.py", line 2152, in visit_enum
bracket-backend   |     self.connection.execute(CreateEnumType(enum))
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1385, in execute
bracket-backend   |     return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/sql/ddl.py", line 80, in _execute_on_connection
bracket-backend   |     return connection._execute_ddl(
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1477, in _execute_ddl
bracket-backend   |     ret = self._execute_context(
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1953, in _execute_context
bracket-backend   |     self._handle_dbapi_exception(
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2134, in _handle_dbapi_exception
bracket-backend   |     util.raise_(
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
bracket-backend   |     raise exception
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
bracket-backend   |     self.dialect.do_execute(
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
bracket-backend   |     cursor.execute(statement, parameters)
bracket-backend   | sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "pg_type_typname_nsp_index"
bracket-backend   | DETAIL:  Key (typname, typnamespace)=(stage_type, 2200) already exists.
bracket-backend   |
bracket-backend   | [SQL: CREATE TYPE stage_type AS ENUM ('SINGLE_ELIMINATION', 'SWISS', 'ROUND_ROBIN')]
bracket-backend   | (Background on this error at: https://sqlalche.me/e/14/gkpj)
bracket-backend   |
bracket-backend   | [2024-02-08 13:06:39 +0000] [10] [ERROR] Application startup failed. Exiting.
bracket-backend   | [2024-02-08 13:06:39 +0000] [10] [INFO] Worker exiting (pid: 10)
bracket-backend   | [2024-02-08 13:06:39 +0000] [1] [ERROR] Worker (pid:10) exited with code 3
bracket-backend   | [2024-02-08 13:06:39 +0000] [9] [INFO] Application startup complete.
bracket-backend   | [2024-02-08 13:06:39 +0000] [9] [ERROR] Traceback (most recent call last):
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/starlette/routing.py", line 747, in lifespan
bracket-backend   |     await receive()
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/uvicorn/lifespan/on.py", line 137, in receive
bracket-backend   |     return await self.receive_queue.get()
bracket-backend   |   File "/usr/local/lib/python3.10/asyncio/queues.py", line 159, in get
bracket-backend   |     await getter
bracket-backend   | asyncio.exceptions.CancelledError
bracket-backend   |
bracket-backend   | [2024-02-08 13:06:39 +0000] [9] [INFO] Worker exiting (pid: 9)
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1953, in _execute_context
bracket-backend   |     self._handle_dbapi_exception(
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2134, in _handle_dbapi_exception
bracket-backend   |     util.raise_(
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
bracket-backend   |     raise exception
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
bracket-backend   |     self.dialect.do_execute(
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
bracket-backend   |     cursor.execute(statement, parameters)
bracket-backend   | sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "pg_type_typname_nsp_index"
bracket-backend   | DETAIL:  Key (typname, typnamespace)=(stage_type, 2200) already exists.
bracket-backend   |
bracket-backend   | [SQL: CREATE TYPE stage_type AS ENUM ('SINGLE_ELIMINATION', 'SWISS', 'ROUND_ROBIN')]
bracket-backend   | (Background on this error at: https://sqlalche.me/e/14/gkpj)
bracket-backend   |
bracket-backend   | [2024-02-08 13:06:39 +0000] [10] [ERROR] Application startup failed. Exiting.
bracket-backend   | [2024-02-08 13:06:39 +0000] [10] [INFO] Worker exiting (pid: 10)
bracket-backend   | [2024-02-08 13:06:39 +0000] [1] [ERROR] Worker (pid:10) exited with code 3
bracket-backend   | [2024-02-08 13:06:39 +0000] [9] [INFO] Application startup complete.
bracket-backend   | [2024-02-08 13:06:39 +0000] [9] [ERROR] Traceback (most recent call last):
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/starlette/routing.py", line 747, in lifespan
bracket-backend   |     await receive()
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/uvicorn/lifespan/on.py", line 137, in receive
bracket-backend   |     return await self.receive_queue.get()
bracket-backend   |   File "/usr/local/lib/python3.10/asyncio/queues.py", line 159, in get
bracket-backend   |     await getter
bracket-backend   | asyncio.exceptions.CancelledError
bracket-backend   |
bracket-backend   | [2024-02-08 13:06:39 +0000] [9] [INFO] Worker exiting (pid: 9)

**Reply by K0styan** [reply]
Seems that there is an error with dbrypt version.
the following lines were found in the docker-compose logs
bracket-backend   | [2024-02-09 14:17:49,676] [bracket] [INFO] Current env: DEVELOPMENT
bracket-backend   | (trapped) error reading bcrypt version
bracket-backend   | Traceback (most recent call last):
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/passlib/handlers/bcrypt.py", line 620, in _load_backend_mixin
bracket-backend   |     version = _bcrypt.__about__.__version__
bracket-backend   | AttributeError: module 'bcrypt' has no attribute '__about__'
bracket-backend   | (trapped) error reading bcrypt version
bracket-backend   | Traceback (most recent call last):
bracket-backend   |   File "/home/bracket/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.10/site-packages/passlib/handlers/bcrypt.py", line 620, in _load_backend_mixin
bracket-backend   |     version = _bcrypt.__about__.__version__
bracket-backend   | AttributeError: module 'bcrypt' has no attribute '__about__'

**Reply by evroon** [reply]
the bcrypt logs are a known problem: https://github.com/pyca/bcrypt/issues/684
but they don't cause Bracket to crash

I created an issue: https://github.com/evroon/bracket/issues/451

**Reply by evroon** [reply]
I can't reproduce your SQL problem

> sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "pg_type_typname_nsp_index"

This likely indicates your db is in a weird state somehow, can you try to reinitialize the db using `./cli.py create-dev-db`?

**Reply by K0styan** [reply]
here you are the message after run reinitialization.
```
root@rxqqlvkoqw:~/bracket/backend# ./cli.py create-dev-db
Traceback (most recent call last):
  File "/root/bracket/backend/./cli.py", line 8, in <module>
    from bracket.config import config
  File "/root/bracket/backend/bracket/config.py", line 5, in <module>
    import sentry_sdk
ModuleNotFoundError: No module named 'sentry_sdk'
```

**Reply by evroon** [reply]
yeah okay you have to be in a pipenv shell, so you can run `pipenv run ./cli.py create-dev-db` instead

**Reply by K0styan** [reply]
here the output of the last comand

root@rxqqlvkoqw:~/bracket/backend# pipenv run ./cli.py create-dev-db
Traceback (most recent call last):
  File "/usr/bin/pipenv", line 33, in <module>
    sys.exit(load_entry_point('pipenv==11.9.0', 'console_scripts', 'pipenv')())
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 696, in main
    with self.make_context(prog_name, args, **extra) as ctx:
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 621, in make_context
    self.parse_args(ctx, args)
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 1018, in parse_args
    rest = Command.parse_args(self, ctx, args)
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 875, in parse_args
    parser = self.make_parser(ctx)
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 821, in make_parser
    for param in self.get_params(ctx):
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 774, in get_params
    help_option = self.get_help_option(ctx)
  File "/usr/lib/python3/dist-packages/pipenv/cli.py", line 26, in get_help_option
    from .import core
  File "/usr/lib/python3/dist-packages/pipenv/core.py", line 21, in <module>
    import requests
  File "/usr/lib/python3/dist-packages/pipenv/vendor/requests/__init__.py", line 65, in <module>
    from . import utils
  File "/usr/lib/python3/dist-packages/pipenv/vendor/requests/utils.py", line 27, in <module>
    from .cookies import RequestsCookieJar, cookiejar_from_dict
  File "/usr/lib/python3/dist-packages/pipenv/vendor/requests/cookies.py", line 172, in <module>
    class RequestsCookieJar(cookielib.CookieJar, collections.MutableMapping):
AttributeError: module 'collections' has no attribute 'MutableMapping'

**Reply by evroon** [reply]
That's not an issue of Bracket, looks like you use some old python version maybe

See: https://stackoverflow.com/a/72607663

**Reply by K0styan** [reply]
I still can not take over this error.

What should be installed on the system before running Brackets?
I tried install Brackets on completely clean system Ubunntu + Python + Docker
I have installed python 3 and Docker 25 with docker compose plugin

**Reply by evroon** [reply]
Do you have Python 3.10 or higher?

And do you use the exact docker-compose from https://github.com/evroon/bracket/blob/master/docker-compose.yml?

**Reply by K0styan** [reply]
I have python 3.10.12

I have copy-pasted this yml file into bracket directory

**Reply by evroon** [reply]
and you run it using `docker-compose up` ?

**Reply by K0styan** [reply]
for run Brackets I use a comand _docker compose up -d_

**Reply by evroon** [reply]
so the error is still the one with MutableMapping when running `pipenv run ./cli.py create-dev-db`?
Do you run that command inside the container and not on the host?
So to run in the container: `sudo docker exec bracket-backend pipenv run ./cli.py create-dev-db`

**Reply by K0styan** [reply]
Please look the attached file. That is what I have done for install Brackets.
[123.zip](https://github.com/evroon/bracket/files/14357036/123.zip)

**Comment by Cyberlytical-LLC** [comment]
Is this going to be fixed? I am still having this issue today and its been months since the last reply. Doesn't look like it ever got fixed.

**Reply by evroon** [reply]
There isn't much I can fix because it seems to be problems with local configurations. If someone can troubleshoot and determine what actually doesn't work, please open an issue describing what's wrong and I will look into it. But for now, it seems to just be local problems caused by the way your environment is set up.

There is also an issue to improve the documentation in this area.

**Reply by Cyberlytical-LLC** [reply]
How can you blame local configs when your instructions literally give 3 steps and it should work? You even admit your documentation is beyond lacking. It doesn't matter where you set the env variables, they don't stick and if you change to a production config you get a postgres error and the API error.

**Reply by evroon** [reply]
> How can you blame local configs when your instructions literally give 3 steps and it should work? 

I said it *seems* to be local configs, I don't blame anyone or anything. It don't know what you expect me to do, I have Bracket running on at least two Ubuntu installations without problems, and one production server as well. I can't SSH into your local environment and troubleshoot the problem myself, nor do I want to.

If you want this problem fixed, I need more information and a clear problem statement. Just saying I got an "internal server error"/"connection refused" error is not enough.

> You even admit your documentation is beyond lacking

It's an free and open source project, I don't owe you anything, not even complete/proper documentation. I work on this in my free time and what you see is what you get. If you want proper documentation or anything else, you can spend some effort and contribute, which I am very open to.
